collaboration-service/controllers/sessionController.js

```diff
@@ -0,0 +1,128 @@
+// controllers/sessionController.js
+
+const codeSessions = {};
+
+const sessionsForUser = {};
+
+// Handle user joining a session
+const joinSession = (socket, io) => {
+  socket.on('join-session', (details) => {
+    const { sessionId, matchedUserId, questionId } = details;
+    const userId = socket.data.user.userId;
+
+    // Input validation
+    if (!sessionId) {
+      return socket.emit('error', 'Session ID is required.');
+    }
+    if (!userId) {
+      return socket.emit('error', 'User ID is required.');
+    }
+    if (!matchedUserId) {
+      return socket.emit('error', 'Matched User ID is required.');
+    }
+    if (userId == matchedUserId) {
+      return socket.emit('error', 'User cannot be matched to him/herself.');
+    }
+    if (!questionId) {
+      return socket.emit('error', 'Question ID is required.');
+    }
+
+    if (!codeSessions[sessionId]) {
+      // Initialize session
+      codeSessions[sessionId] = {
+        code: '',
+        users: new Set([userId, matchedUserId]),
+        activeConnections: 0,
+        questionId: questionId
+      };
+    } else if (
+      !codeSessions[sessionId].users.has(userId) ||
+      !codeSessions[sessionId].users.has(matchedUserId)
+    ) {
+      return socket.emit('error', 'Invalid user/match Id for this session ID.');
+    } else if (codeSessions[sessionId].questionId != questionId) {
+      return socket.emit('error', 'Invalid question Id for this session ID.');
+    }
+
+    socket.join(sessionId);
+
+    codeSessions[sessionId].activeConnections += 1;
+    delete codeSessions[sessionId].timeoutStart;
```

Comment:
Do we need a check here for `if (activeConnections >= 2)`? to prevent more than 2 users connecting

Reply:
good spot! added the check